### PR TITLE
Stabilize unsigned and float variants of `num_midpoint` feature

### PR DESCRIPTION
**Pull request by Urgau**
This PR proposes that we stabilize the unsigned variants of the [`num_midpoint`](https://github.com/rust-lang/rust/issues/110840#issue-1684506201) feature as well as the floats variants, since they are not subject to any unresolved questions, which is equivalent to doing `(a + b) / 2` (and `(a + b) >> 1`) in a sufficiently large number.

The stabilized API surface would be:

```rust
/// Calculates the middle point of `self` and `rhs`.
///
/// `midpoint(a, b)` is `(a + b) / 2` as if it were performed in a sufficiently-large unsigned integral type.
/// This implies that the result is always rounded towards negative infinity and that no overflow will ever occur.

impl u{8,16,32,64,128,size} {
    pub const fn midpoint(self, rhs: Self) -> Self;
}

impl NonZeroU{8,16,32,64,size} {
    pub const fn midpoint(self, rhs: Self) -> Self;
}

impl f{32,64} {
    pub const fn midpoint(self, rhs: Self) -> Self;
}
```

The signed variants `u{8,16,32,64,128,size}` would remain gated, until a decision is made about the rounding mode, in other words that the [unresolved questions](https://github.com/rust-lang/rust/issues/110840#issue-1684506201) are resolved.

cc @rust-lang/libs-api
cc @scottmcm
r? libs-api

### CODE REVIEW

**Review by dtolnay — APPROVED** [review]

**Comment by dtolnay** [comment]
@rfcbot fcp merge

**Comment by tgross35** [comment]
Float `midpoint` should probably get a `const` gate, either here or separately. Bit unfortunate that it doesn't already have one to get stabilized at the same time.

**Comment by Urgau** [comment]
> Float `midpoint` should probably get a `const` gate, either here or separately.

Good point, updated the PR to include the const stabilization of `f{32,64}::midpoint` since there is nothing preventing it.

**Comment by joshtriplett** [comment]
> That said, I feel obliged to mention that in the tracking issue that some people want `midpoint(5, 10) == 7` but `midpoint(10, 5) == 8`, where it's consistent with `a + (b - a) / 2` instead, so wanted to make sure that libs-api discusses the rounding explicitly.

I would advocate the behavior that `a.midpoint(b) == b.midpoint(a)` for all `a` and `b`.

File: library/core/src/num/f128.rs [hunk 1]
```diff
@@ -829,8 +828,8 @@ impl f128 {
     /// ```
     #[inline]
     #[unstable(feature = "f128", issue = "116909")]
-    // #[unstable(feature = "num_midpoint", issue = "110840")]
-    pub fn midpoint(self, other: f128) -> f128 {
+    #[rustc_const_unstable(feature = "f128", issue = "116909")]
+    pub const fn midpoint(self, other: f128) -> f128 {
```
Comment (tgross35) [inline library/core/src/num/f128.rs:832]: I believe you don't need a separate `rustc_const_unstable` gate anymore if the whole thing is `unstable`, just the `const` keyword (applies to `f16` too)

**Comment by RalfJung** [comment]
Failed in PR CI in https://github.com/rust-lang/rust/pull/133682#issuecomment-2509128744. Probably needs a  rebase.

@bors r-

File: library/core/src/num/f128.rs [hunk 2]
```diff
@@ -829,8 +828,7 @@ impl f128 {
     /// ```
     #[inline]
     #[unstable(feature = "f128", issue = "116909")]
-    // #[unstable(feature = "num_midpoint", issue = "110840")]
-    pub fn midpoint(self, other: f128) -> f128 {
+    pub const fn midpoint(self, other: f128) -> f128 {
```
Comment (RalfJung) [inline library/core/src/num/f128.rs:831]: You added a `const` here, you will also have to add a `rustc_const_unstable`.
```suggestion
    #[rustc_const_unstable(feature = "f128", issue = "116909")]
    pub const fn midpoint(self, other: f128) -> f128 {
```
Same for the f16 function.

File: library/core/src/num/f16.rs [hunk 3]
```diff
@@ -804,8 +803,8 @@ impl f16 {
     /// ```
     #[inline]
     #[unstable(feature = "f16", issue = "116909")]
-    // #[unstable(feature = "num_midpoint", issue = "110840")]
-    pub fn midpoint(self, other: f16) -> f16 {
+    #[rustc_const_unstable(feature = "f128", issue = "116909")]
```
[the remark this answers is not in the document]
Comment (RalfJung) [reply library/core/src/num/f16.rs:806]: *oops* right, good point.

@Urgau can you submit a follow-up fixing that?

[the remark this answers is not in the document]
Comment (Urgau) [reply library/core/src/num/f16.rs:806]: My bad, fixed in https://github.com/rust-lang/rust/pull/133763.

**Pull request merged**